org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java

```diff
@@ -319,15 +320,62 @@ protected void incrementalBuild(IResourceDelta delta, final IProgressMonitor mon
 		progress.worked(2);
 		Task task = closedProjectsQueue.exhaust();
 		try {
-			addInfosFromTask(task, toBeBuilt);
-			doBuild(toBeBuilt, progress.split(8), BuildType.INCREMENTAL);
+			addInfosFromTaskAndBuild(task, toBeBuilt, BuildType.INCREMENTAL, progress.split(8));
 		} catch(Exception e) {
 			task.reschedule();
 			throw e;
 		}
 	}
 
-	private void addInfosFromTask(Task task, final ToBeBuilt toBeBuilt) {
+	private static Boolean mustCallDeprecatedDoBuild = null;
+	private static boolean wasDeprecationWarningLogged = false;
+	
+	/**
+	 * @since 2.18
+	 */
+	protected void addInfosFromTaskAndBuild(Task task, final ToBeBuilt toBeBuilt, BuildType buildType, IProgressMonitor monitor)
+			throws CoreException {
```

Comment:
highly sophisticated :)

Reply:
I tried to break only few things.

What I was also thinking about: changing the constructor of BuildData to smth like `BuildData(Map<Key<T>, T>)` to become more forward compatible.